srml/contract/src/wasm/runtime.rs

```diff
@@ -610,21 +628,47 @@ define_env!(Env, <E: Ext>,
 		Ok(())
 	},
 
-	// Deposit a contract event with the data buffer.
-	ext_deposit_event(ctx, data_ptr: u32, data_len: u32) => {
+	// Deposit a contract event with the data buffer and optional list of topics. There is a limit
+	// on the maximum number of topics specified by `max_event_topics`.
+	//
+	// - topics_ptr - a pointer to the buffer of topics encoded as `Vec<T::Hash>`. The value of this
+	//   is ignored if `topics_len` is set to 0. The topics list can't contain duplicates.
+	// - topics_len - the length of the topics buffer. Pass 0 if you want to pass an empty vector.
+	// - data_ptr - a pointer to a raw data buffer which will saved along the event.
+	// - data_len - the length of the data buffer.
+	ext_deposit_event(ctx, topics_ptr: u32, topics_len: u32, data_ptr: u32, data_len: u32) => {
+		let topics = match topics_len {
+			0 => Vec::new(),
+			_ => {
+				let topics_buf = read_sandbox_memory(ctx, topics_ptr, topics_len)?;
+				Vec::<TopicOf<<E as Ext>::T>>::decode(&mut &topics_buf[..])
+					.ok_or_else(|| sandbox::HostError)?
+			}
+		};
+
+		// If there are more than `max_event_topics`, then trap.
+		if topics.len() > ctx.schedule.max_event_topics as usize {
+			return Err(sandbox::HostError);
+		}
+
+		// Check for duplicate topics. If there are any, then trap.
+		if (1..topics.len()).any(|i| topics[i..].contains(&topics[i - 1])) {
```

Comment:
If the order of the topics does not matter, we can do this in O(n log n) by sorting first.

Reply:
The order of topics doesn't matter indeed. I initially approached with `sort_unstable` (`sort_unstable` because it doesn't require additional memory), but alas sorting requires `T::Hash` to be `Ord` which I was not sure about.

I expect `N` here would be really small (say 10 at most) so I assumed that quadratic will work here on par with sorting.

Reply:
Good point @thiolliere ! Well I think it is worth to leverage the sorting approach then!

Reply:
Fixed!